commons/src/main/java/org/infinispan/counter/util/EncodeUtil.java

```diff
@@ -0,0 +1,160 @@
+package org.infinispan.counter.util;
+
+import java.util.function.Consumer;
+import java.util.function.IntConsumer;
+import java.util.function.IntSupplier;
+import java.util.function.LongConsumer;
+import java.util.function.LongSupplier;
+import java.util.function.Supplier;
+
+import org.infinispan.commons.logging.Log;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.counter.api.CounterConfiguration;
+import org.infinispan.counter.api.CounterType;
+import org.infinispan.counter.api.Storage;
+
+/**
+ * Utility class to handle encoding or decoding counter's classes.
+ *
+ * @author Pedro Ruivo
+ * @since 9.2
+ */
+public final class EncodeUtil {
+
+   private static final Log log = LogFactory.getLog(EncodeUtil.class, Log.class);
+
+   /*
+   00000000
+        ||^- 1 WEAK, 0 STRONG
+        |^-- 1 BOUNDED, 0 UNBOUNDED
+        ^--- 1 PERSISTENT, 0 VOLATILE
+    */
+   private static final byte WEAK_COUNTER = 0x01;
+   private static final byte BOUNDED_COUNTER = 0x02;
+   private static final byte UNBOUNDED_COUNTER = 0x00;
+
+   private EncodeUtil() {
+   }
+
+   /**
+    * Decodes the {@link CounterType}.
+    *
+    * @return the decoded {@link CounterType}.
+    * @see #encodeTypeAndStorage(CounterConfiguration)
+    */
+   public static Storage decodeStorage(byte flags) {
+      return (flags & 0x04) == 0 ? Storage.VOLATILE : Storage.PERSISTENT;
+   }
+
+   /**
+    * Decodes the {@link Storage}.
+    *
+    * @return the decoded {@link Storage}.
+    * @see #encodeTypeAndStorage(CounterConfiguration)
+    */
+   public static CounterType decodeType(byte flags) {
+      switch (flags & 0x03) {
+         case WEAK_COUNTER:
+            return CounterType.WEAK;
+         case BOUNDED_COUNTER:
+            return CounterType.BOUNDED_STRONG;
+         case UNBOUNDED_COUNTER:
+            return CounterType.UNBOUNDED_STRONG;
+         default:
+            throw log.invalidCounterTypeEncoded();
+      }
+   }
+
+   private static byte encodeStorage(Storage storage) {
+      switch (storage) {
+         case VOLATILE:
+            return 0x00;
+         case PERSISTENT:
+            return 0x04;
+         default:
+            throw new IllegalStateException("Unknown storage mode: " + storage);
+      }
+   }
+
+   private static byte encodeType(CounterType type) {
+      switch (type) {
+         case UNBOUNDED_STRONG:
+            return UNBOUNDED_COUNTER;
+         case BOUNDED_STRONG:
+            return BOUNDED_COUNTER;
+         case WEAK:
+            return WEAK_COUNTER;
+         default:
+            throw new IllegalStateException();
+      }
+   }
+
+   /**
+    * Encodes the {@link Storage} and the {@link CounterType}.
+    * <p>
+    * See the documentation for further details about the encoding.
+    * <p>
+    * TODO! link to docs
```

Comment:
todo

Reply:
there is no docs yet :)

Reply:
I've added a link to where the docs should be :)